src/sourceParser.js

```diff
@@ -7,13 +7,53 @@
 
 const importParser = /\bimport\s*(?:\(|\/[/*])/;
 
-export function rejectImportExpressions(s) {
-  const index = s.search(importParser);
-  if (index !== -1) {
-    // todo: if we have a full parser available, use it here. If there is no
-    // 'import' token in the string, we're safe.
-    // if (!parse(s).includes('import')) return;
-    const linenum = s.slice(0, index).split('\n').length; // more or less
-    throw new SyntaxError(`possible import expression rejected around line ${linenum}`);
+export function parseSource(s, options) {
+  const { infixBangResolver } = Object(options);
+
+  // eslint-disable-next-line global-require
+  const parser = require('@agoric/babel-parser');
+  if (infixBangResolver) {
+    return parser.parse(s, {
+      plugins: [['infixBang', { resolver: infixBangResolver }]]
+    });
+  }
+  return parser.parse(s);
+}
+
+export function generateSource(s, ast) {
+  // eslint-disable-next-line global-require
+  const { default: generate } = require('@babel/generator');
+  const { code } = generate(ast, {}, s);
+  return code;
+}
+
+export function rejectImportExpressions(s, ast) {
+  if (!ast) {
+    // We didn't yet do a full parse, so try the cheap regexp first.
+    const index = s.search(importParser);
+    if (index === -1) {
+      // No possible import expression.
+      return;
+    }
+
+    // if we have a full parser available, use it here. Below we check the
+    // resulting ast.
+    try {
+      ast = parseSource(s);
+    } catch (e) {
+      const linenum = s.slice(0, index).split('\n').length; // more or less
+      throw new SyntaxError(`possible import expression rejected around line ${linenum}: ${e}`);
+    }
   }
+
+  // We have a parse, so traverse it.
```

Comment:
This looks like you may be rejecting assuming an accurate parse, even when you're not regenerating from ast. Is that so?